## What is Pandas Profiling or ydata Profiling ?

Pandas Profiling automatically generates an exploratory data analysis (EDA) report from a pandas DataFrame.

## Installation : 

In [8]:
!pip install ydata-profiling

### 1. Import libraries 

In [9]:
from ydata_profiling import ProfileReport
import pandas as pd

### 2. Load your data

In [25]:
df = pd.read_csv('ford.csv')
df.shape

(17966, 9)

### 3. Generate the profile

In [22]:
profile = ProfileReport(df, title="Ford Report",explorative=True)


# Use explorative=True when:
# Dataset is small/medium
# You are in the early EDA phase

### 4. View the report

In [ ]:
profile.to_file("report.html")

# open report.html in browser to see the report

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

100%|██████████| 9/9 [00:00<00:00, 322.65it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## Analyse the Data :

### Dataset Overview (ALWAYS start here)

- Where in report: Overview section


`Check:`
- Number of rows & columns
- Missing cells (%)
- Duplicate rows
- Memory usage


Why it matters:

- Tells you how “healthy” your dataset is
- Helps you decide:
- Can I load it fully?
- Do I need cleaning first?


In this report : 
- Missing cells : 	0
- Missing cells (%) : 	0.0%
- Duplicate rows : 	148
- Duplicate rows (%) : 	0.8%
- Total size in memory :	3.7 MiB

## 2. Data Types (Very Important)

- Variables → each column

`Check:`

- Are data types correct?
- Numbers as numbers?
- Dates as datetime?
- Categories as categorical?

`Common issues:`

- age stored as string
- date stored as object
- salary contains symbols ($, ,)

Fix early, or analysis & models will break.


In this report : 

- It has in total 17,966 rows and 9 columns
- model : categorical 
- year : real number (int)
- price : real number (int)
- transmission : categorical
- mileage : real number (int)
- fuelType : categorical
- tax : real number (int)
- mpg : real number (float) 
- engineSize : real number (float)  

-> All Ok




**A column is labeled *Real number* when:**

- Its dtype is typically float (float32, float64)
- Or it contains decimal values even if stored as numeric
- And it’s not detected as categorical (low unique values, repeated patterns, etc.)

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17966 entries, 0 to 17965
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         17966 non-null  object 
 1   year          17966 non-null  int64  
 2   price         17966 non-null  int64  
 3   transmission  17966 non-null  object 
 4   mileage       17966 non-null  int64  
 5   fuelType      17966 non-null  object 
 6   tax           17966 non-null  int64  
 7   mpg           17966 non-null  float64
 8   engineSize    17966 non-null  float64
dtypes: float64(2), int64(4), object(3)
memory usage: 1.2+ MB


### Missing Values Analysis

- Overview + Variables section

Check:

- Which columns have missing values?
- How many?
- Pattern of missingness (random or not?)
- Why it matters:
- Missing data can bias your model


Some columns may need:
- Imputation
- Removal
- Special handling

*🚨 Rule of thumb*

- `40% missing → consider dropping the column`
- Critical column missing → investigate deeply




In this report : 

- there is no missing values

### Duplicates & Unique Values

- Overview + Variables

Check:
- Duplicate rows
- Columns with:
    - Only one value (constant)
    - Too many unique values (IDs)


Why it matters:
- Duplicates distort statistics
- IDs should not be used as features

*🚨 Warnings you’ll see :*
- “Constant column”
- “High cardinality”


In this report : 
- Duplicate rows: 148 (0.8%)
- tax column has : 
    - Zeros	    :   2153
    - Zeros (%) :	12.0%
- Transmission (Highly Imbalanced)
    - Manual: 86%
    - Automatic + Semi-Auto: 14%
- Fuel Type (Severely Imbalanced)
    - Petrol + Diesel ≈ 99.8%
    - Hybrid / Electric ≈ 0.2%


### 5️. Distribution of Numerical Columns

- Variables → Numeric columns

Check:
- Mean vs Median
- Min / Max
- Histogram shape

Look for:
- Skewed distributions
- Extreme values (outliers)
- Impossible values (age = -5)


*In this report Overall Observations :*

- Extreme or impossible values:
    - year = 2060
    - mpg = 201.8
    - engineSize = 0
    - mileage = 1 (maybe acceptable if brand new)

- Strongly skewed numeric columns:
    - mileage, price, engineSize → may need log-transform
    - Right-skewed → mean > median
    - Tight distributions / outliers:
    - year mostly 2016–2018, except extreme max
    - tax mostly 145

- Action plan for cleaning before analysis/ML:
    - Remove/fix impossible values (year > 2026, engineSize = 0, mpg > 120?)
    - Consider log-transform for skewed columns (price, mileage, engineSize)
    - Check min values that are suspicious (price = 495, mileage = 1)


*Alerts :*
- Dataset has 148 (0.8%) duplicate rows	                - Duplicates
- engineSize is highly overall correlated with model	-  High correlation
- mileage is highly overall correlated with price and 1 other fields 	High correlation
- model is highly overall correlated with engineSize    -	High correlation
- price is highly overall correlated with mileage and 1 other fields   - High correlation
- year is highly overall correlated with mileage and 1 other fields	   - High correlation
- transmission is highly imbalanced (55.2%)	    -  Imbalance
- fuelType is highly imbalanced (60.4%)	    -    Imbalance
- tax has 2153 (12.0%) zeros    - 	Zeros
